### PR TITLE
refac: improve event data parsing and structuring

### PR DESCRIPTION
**Pull request by Akshola00**
# Improve Event Data Parsing and Structuring

### Fixes
This PR addresses Issue #234 by improving the data parsing and structuring for on-chain event data. Specifically:
1. Parses on-chain event data to a human-readable format.
2. Replaces index-based data access with named fields through a serializer.
3. Adds a `RepaymentEventSerializer` model using Pydantic in `serializers.py`.

### Summary of Changes

1. **Parsing Data in `parse_repayment_event`**:
   - Updated `parse_repayment_event` to convert raw event data to named attributes, transforming `data[0]` to `data.user` for improved readability and accessibility.

2. **Serializer Model in `serializers.py`**:
   - Created a new `RepaymentEventSerializer` class using Pydantic. This model handles the structured event data:

3. **Refactoring `process_repayment_event`**:
   - Updated `process_repayment_event` to use the `RepaymentEventSerializer` model, replacing index-based access:

### CODE REVIEW

File: data_handler/handler_tools/data_parser/serializers.py [hunk 1]
```diff
@@ -86,3 +86,22 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
         if not value.isdigit():
             raise ValueError("%s field is not numeric" % info.field_name)
         return Decimal(str(int(value, base=16)))
+
+class RepaymentEventData(BaseModel):
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:90]: add docstring

File: data_handler/handler_tools/data_parser/serializers.py [hunk 2]
```diff
@@ -86,3 +86,22 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
...
+    repayer: str
+    beneficiary: str
+    token: str
+    raw_amount: str
+    face_amount: str
+
+class RepaymentEventSerializer(BaseModel):
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:97]: add docstring

File: data_handler/handler_tools/data_parser/serializers.py [hunk 3]
```diff
@@ -86,3 +86,22 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
...
+    repayer: Any       # Replace `Any` with a more specific type if known (e.g., `str` or `int`)
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:98]: remove comment from CHAT GPT, these comments are useless and it's just recommendation.

File: data_handler/handler_tools/data_parser/serializers.py [hunk 4]
```diff
@@ -86,3 +86,22 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
...
+    beneficiary: Any   # Same here if you know the exact type
+    token: Any         # Again, replace with the specific type if known
+    raw_amount: str    # Representing this as a string, as it will be parsed as hexadecimal
+    face_amount: str   # Same as above for consistency
+    block_number: int
+    timestamp: int
+
+    class Config:
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:106]: add docstring

File: data_handler/handler_tools/data_parser/serializers.py [hunk 5]
```diff
@@ -86,3 +86,22 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
...
+        arbitrary_types_allowed = True  # Allows flexibility for non-standard types if needed
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:107]: Remove comment from CHAT GPT

File: data_handler/handler_tools/data_parser/zklend.py [hunk 6]
```diff
@@ -59,10 +59,26 @@ def parse_borrowing_event(cls, event_data):
         # TODO: Implement parsing logic for Borrowing event
         pass
 
-    @classmethod
-    def parse_repayment_event(cls, event_data):
-        # TODO: Implement parsing logic for Repayment event
-        pass
+@classmethod
+def parse_repayment_event(cls, event_data):
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/zklend.py:63]: add typing to param and return type

File: data_handler/handlers/loan_states/zklend/events.py [hunk 7]
```diff
@@ -303,27 +305,39 @@ def process_borrowing_event(self, event: pd.Series) -> None:
...
-        self.loan_entities[user].extra_info.block = event["block_number"]
-        self.loan_entities[user].extra_info.timestamp = event["timestamp"]
-        if user == self.verbose_user:
-            logging.info(
-                "In block number = {}, raw amount = {} of token = {} was repayed.".format(
-                    event["block_number"],
-                    raw_amount,
-                    token,
-                )
+def process_repayment_event(self, event: pd.Series) -> None:
```
Comment (djeck1432) [inline data_handler/handlers/loan_states/zklend/events.py:309]: fix spaces, you broke the whole class, 
it should have 
```
    def process_repayment_event(self, event: pd.Series) -> None: # RIGHT 
def process_repayment_event(self, event: pd.Series) -> None: # WRONG
```

File: data_handler/handlers/loan_states/zklend/events.py [hunk 8]
```diff
@@ -303,27 +305,39 @@ def process_borrowing_event(self, event: pd.Series) -> None:
...
+        "beneficiary": event["data"][1],
+        "token": event["data"][2],
+        "raw_amount": event["data"][3],
+        "face_amount": event["data"][4],
+        "block_number": event["block_number"],
+        "timestamp": event["timestamp"]
+    })
+    
+    # Now we access `beneficiary`, `token`, `raw_amount` and others by name.
+    user = add_leading_zeros(parsed_event.beneficiary)
```
Comment (djeck1432) [inline data_handler/handlers/loan_states/zklend/events.py:322]: move `add_leading_zeros` to `field_validation`

File: data_handler/handlers/loan_states/zklend/events.py [hunk 9]
```diff
@@ -303,27 +305,39 @@ def process_borrowing_event(self, event: pd.Series) -> None:
...
-        if user == self.verbose_user:
-            logging.info(
-                "In block number = {}, raw amount = {} of token = {} was repayed.".format(
-                    event["block_number"],
-                    raw_amount,
-                    token,
-                )
+def process_repayment_event(self, event: pd.Series) -> None:
+    # Parse event data into our RepaymentEventSerializer model to access by named attributes
+    parsed_event = RepaymentEventSerializer.parse_obj({
```
Comment (djeck1432) [inline data_handler/handlers/loan_states/zklend/events.py:311]: insert only event, and parse it in `Serializer` not here.

**Review by djeck1432 — CHANGES_REQUESTED** [review]

File: data_handler/handlers/loan_states/zklend/events.py (hunk 9 above)
Comment (Akshola00) [reply data_handler/handlers/loan_states/zklend/events.py:311]: On it now

Comment (djeck1432) [reply data_handler/handlers/loan_states/zklend/events.py:311]: @Akshola00 Ping me once you are done, I will be here 2-3 hours

File: data_handler/handler_tools/data_parser/serializers.py (hunk 2 above)
Comment (Akshola00) [reply data_handler/handler_tools/data_parser/serializers.py:107]: done

File: data_handler/handler_tools/data_parser/serializers.py (hunk 3 above)
Comment (Akshola00) [reply data_handler/handler_tools/data_parser/serializers.py]: done

File: data_handler/handler_tools/data_parser/serializers.py (hunk 4 above)
Comment (Akshola00) [reply data_handler/handler_tools/data_parser/serializers.py:168]: done

File: data_handler/handler_tools/data_parser/serializers.py (hunk 5 above)
Comment (Akshola00) [reply data_handler/handler_tools/data_parser/serializers.py]: done

File: data_handler/handlers/loan_states/zklend/events.py (hunk 7 above)
Comment (Akshola00) [reply data_handler/handlers/loan_states/zklend/events.py]: done

File: data_handler/handlers/loan_states/zklend/events.py (hunk 8 above)
Comment (Akshola00) [reply data_handler/handlers/loan_states/zklend/events.py]: done

File: data_handler/handlers/loan_states/zklend/events.py (hunk 9 above)
Comment (Akshola00) [reply data_handler/handlers/loan_states/zklend/events.py]: done

**Comment by Akshola00** [comment]
@djeck1432 made all changes

File: data_handler/handler_tools/data_parser/serializers.py [hunk 10]
```diff
@@ -83,7 +83,101 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
...
+        raw_amount (str): The raw amount of the repayment as provided, before any conversions or calculations.
+        face_amount (str): The face amount of the repayment, representing the value after necessary conversions.
+    """
+    repayer: str
+    beneficiary: str
+    token: str
+    raw_amount: str
+    face_amount: str
+
+class RepaymentEventSerializer(BaseModel):
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:108]: why do we need this class if can have only `RepaymentEventData`?
Please redudant it and move all logic there to `RepaymentEventData`

File: data_handler/handler_tools/data_parser/serializers.py [hunk 11]
```diff
@@ -83,7 +83,101 @@ def validate_valid_numbers(cls, value: str, info: ValidationInfo) -> Decimal:
...
+    """
+    repayer: Any    
+    beneficiary: Any
+    token: Any      
+    raw_amount: str 
+    face_amount: str
+    block_number: int
+    timestamp: int
+
+    @validator("beneficiary", "token", pre=True, always=True)
```
Comment (djeck1432) [inline data_handler/handler_tools/data_parser/serializers.py:134]: `validator` decorator is depricated use `field_validator`

**Review by djeck1432 — CHANGES_REQUESTED** [review]

**Comment by Akshola00** [comment]
@djeck1432 changes resolved

File: data_handler/handlers/loan_states/zklend/events.py [hunk 12]
```diff
@@ -303,23 +305,23 @@ def process_borrowing_event(self, event: pd.Series) -> None:
...
-        # `face_amount`.
-        # Example: https://starkscan.co/event/0x06fa3dd6e12c9a66aeacd2eefa5a2ff2915dd1bb4207596de29bd0e8cdeeae66_5.
-        user = add_leading_zeros(event["data"][1])
-        token = add_leading_zeros(event["data"][2])
+        parsed_event = RepaymentEventSerializer.parse_event(event)
 
-        raw_amount = decimal.Decimal(str(int(event["data"][3], base=16)))
+        user = parsed_event.beneficiary
+        token = parsed_event.token
+        raw_amount = decimal.Decimal(str(int(parsed_event.raw_amount, base=16)))
```
Comment (djeck1432) [inline data_handler/handlers/loan_states/zklend/events.py:314]: `decimal.Decimal(str(int(parsed_event.raw_amount, base=16)))` move this logic to `RepaymentEventSerializer`

**Review by djeck1432 — CHANGES_REQUESTED** [review]
Please sync with `master`

**Comment by djeck1432** [comment]
@Akshola00 Pleae fix comments and resolve conflicts

**Comment by Akshola00** [comment]
Will do so now

**Comment by Akshola00** [comment]
@djeck1432 resloved confilcts

**Comment by djeck1432** [comment]
@Akshola00 Please, sync with `master` I'd like to merge it

**Comment by Akshola00** [comment]
sorry about that 
ive synced with master now @djeck1432

**Review by djeck1432 — APPROVED** [review]
Looks good to me

**Pull request merged**